datafusion/core/src/datasource/physical_plan/parquet/statistics.rs

```diff
@@ -1041,21 +1043,20 @@ impl<'a> StatisticsConverter<'a> {
     pub fn data_page_row_counts<I>(
         &self,
         column_offset_index: &ParquetOffsetIndex,
-        row_group_metadatas: &[RowGroupMetaData],
+        row_group_metadatas: &'a [RowGroupMetaData],
         row_group_indices: I,
-    ) -> Result<ArrayRef>
+    ) -> Result<UInt64Array>
     where
         I: IntoIterator<Item = &'a usize>,
     {
-        let data_type = self.arrow_field.data_type();
-
         let Some(parquet_index) = self.parquet_index else {
-            return Ok(self.make_null_array(data_type, row_group_indices));
+            // no matching column found in parquet_index;
+            // thus we cannot extract page_locations in order to determine
+            // the row count on a per DataPage basis.
+            // We use `row_group_row_counts` instead.
+            return Self::row_group_row_counts(row_group_metadatas);
```

Comment:
I see -- this is a tricky situation where there is no column and thus no information on data pages. 

Another potential behavior that might make sense here would be to return an error because unlike other functions in `StatisticsConverter` there is no way to "gracefully" ignore missing information

Or we could possible return an array with zero rows 🤔

Reply:
...perhaps looking at the behavior of `row_group_row_count` might help.

My main question here is: 
Why do we return row counts for a non existing column in `row_group_row_counts`?

A column that does not exists cannot have valid rows, or a row count in that regard? So returning a null_array would indicate those missing information? Suppose we had just a single column, and for whatever reason, it does not exist in the parquet file (or does not match). With the current implementation we would still return a 'valid' row count in this scenario, accessing a somewhat invalid / useless parquet file? 

I'm not sure one can follow my train of thought here, however I think we should return a null_array in both cases for `row_groups` and `data_pages` as well.

But, I might be missing something. Perhaps you can explain the reasoning why we don't return a null_array in the current `row_group_row_counts` impl @alamb? That would be nice and might help guide our decision here.

Reply:
> Why do we return row counts for a non existing column in row_group_row_counts?

I think the (not great) reason is that it is the API needed for PruningStatistics here: https://github.com/apache/datafusion/blob/a923c659cf932f6369f2d5257e5b99128b67091a/datafusion/core/src/datasource/physical_plan/parquet/row_groups.rs#L386-L390

Also, since the `ParquetMetadata` knows how many row groups there are even when there are no row group statistics, it is possible

For data pages, it is different as if the "page index" is not present then I don't think there is any way to know how many data pages there are other than reading the file 

The more we explore this, the more you have convinced me that we shouldn't return row counts for non existent columns in `row_group_row_counts` either

So I guess my new proposal would be to return `Option` like:

```rust
impl<'a> StatisticsConverter<'a> {
...
  /// return OK(None) if the column does not exist
  pub(crate) fn null_counts_page_statistics<'a, I>(iterator: I) -> Result<Option<UInt64Array>> {
  ...
  }

  /// return OK(None) if the column does not exist
    pub fn data_page_row_counts<I>(
        &self,
        column_offset_index: &ParquetOffsetIndex,
        row_group_metadatas: &[RowGroupMetaData],
        row_group_indices: I,
    ) -> Result<Option<UInt64Array>>
    where
        I: IntoIterator<Item = &'a usize>,
    {
    ...
    }

}
```

The rationale to return an Option rather than an error is that creating and ignoring `DataFusionError` via `ok()` still requires a string allocation, which is wasteful

I realize this is done many places already in the statistics extraction code, but I think for those cases it is meant to make the code resilent to incorrectly encoded parquet files rather than something that is "expected" to happen

Reply:
> The more we explore this, the more you have convinced me that we shouldn't return row counts for non existent columns in `row_group_row_counts` either

But I'll guess this is a topic (or PR) for another day?

So for now, we only change `data_page_row_counts` and merge this? I'd be fine with that.

Reply:
> So for now, we only change data_page_row_counts and merge this? I'd be fine with that.

I think that is a good idea and I will file a ticket to fix `row_group_row_counts` in a follow on

Reply:
I changed it accordingly. Thanks again for your input here

Reply:
Filed https://github.com/apache/datafusion/issues/10965 to track the changes to row_group_row_counts